### PR TITLE
feat: added min_p option from Request to CUDA logit processing layer

### PR DESCRIPTION
**Pull request by DwieDave**

### CODE REVIEW

File: llgtrt/src/routes/completions.rs [hunk 1]
```diff
@@ -432,11 +433,23 @@ async fn mk_req_info(
         }
         r
     } else {
-        if req_params.temperature < 0.00001 {
-            req_params.temperature = 0.0;
-            req_params.top_k = 1;
+        if params.min_p.is_finite() && params.min_p > 0.0 {
```
Comment (user-0a) [inline llgtrt/src/routes/completions.rs]: instead of implementing this logic like so, it may be cleaner to update the `llg_grammar` function to return the dummy constraint if `min_p` exists (and there isn't already a non-dummy constraint being enforced).

with the above code change, this extra codepath wouldn't be needed since it would just trigger the `Some(grm) = llg_grammar(params)` codepath above.

File: llgtrt/src/routes/openai.rs [hunk 2]
```diff
@@ -258,6 +261,10 @@ fn default_min_tokens() -> usize {
     1
 }
 
+fn default_min_p() -> f32 {
+    f32::NEG_INFINITY
```
Comment (user-0a) [inline llgtrt/src/routes/openai.rs]: i think this should maybe be 0.

the "default" value of ln_min_p should be -inf, but this corresponds to a default value for min_p of 0

File: trtllm-c/logits.cpp [hunk 3]
```diff
@@ -130,8 +131,10 @@ static void logitsPostProcessorFn(std::vector<tle::IdType> const& reqIds, std::v
     int64_t* logitPtrs = (int64_t*) masksData;
     int64_t* masksOffsets = logitPtrs + batchSize;
     float* temperatures = (float*) (logitPtrs + 2 * batchSize);
+    float* ln_min_ps = temperatures + batchSize;
```
Comment (user-0a) [inline trtllm-c/logits.cpp]: perhaps more canonical to write `(float*) (logitPtrs + 3 * batchSize);` here, so it matches up with `temperatures`

Comment (user-0a) [reply trtllm-c/logits.cpp]: additionally, i think this code overwrites the `tlc_mask_fraction_ptr` which resides at `masksData + max_batch_size * sizeof(int64_t) * 3`.

i think you will need to change `hd_size` from `max_batch_size * sizeof(int64_t) * 4` to `max_batch_size * sizeof(int64_t) * 5` to make more room overall, then change `tlc_mask_fraction_ptr` from `masksData + max_batch_size * sizeof(int64_t) * 3` to `masksData + max_batch_size * sizeof(int64_t) * 4`.

these changes will allow the `ln_min_ps` to sit at `(float*) (logitPtrs + 3 * batchSize)` without conflicting with potential masks

File: llgtrt/src/routes/openai.rs (hunk 2 above)
Comment (user-0a) [reply llgtrt/src/routes/openai.rs]: note that this change will have downstream effects on that various parts of the code that check `isfinite`, etc.

File: llgtrt/src/routes/openai.rs [hunk 4]
```diff
@@ -165,6 +165,9 @@ pub struct CommonCreateParams {
     /// tokens comprising the top 10% probability mass are considered.
     #[serde(default = "default_top_p")]
     pub top_p: f32,
+    /// The minimum log probability of the generated completion.
```
Comment (user-0a) [inline llgtrt/src/routes/openai.rs]: Ah, it's a bit confusing but min_p is just a multiplier, not the probability itself. I think slightly more accurate would be

/// Filters out tokens with probability less than min_p multiplied by the probability of the most likely token

File: trtllm-c/logits.cpp [hunk 5]
```diff
@@ -130,8 +131,10 @@ static void logitsPostProcessorFn(std::vector<tle::IdType> const& reqIds, std::v
     int64_t* logitPtrs = (int64_t*) masksData;
     int64_t* masksOffsets = logitPtrs + batchSize;
     float* temperatures = (float*) (logitPtrs + 2 * batchSize);
+    float* ln_min_ps = (float*) (logitPtrs + 3 * batchSize);
```
Comment (user-0a) [inline trtllm-c/logits.cpp]: supernit: it seems that only the variables passed to the C side are snake case. e.g. `maskOffsets` and `temperatures` directly above this variable use pascal case over snake case. maybe can change this to `lnMinPs` to follow suit.

File: trtllm-c/tlc.h [hunk 6]
```diff
@@ -120,6 +122,7 @@ extern "C"
         float presence_penalty;
         float priority;
         uint32_t top_k;
+        float min_p;
```
Comment (user-0a) [inline trtllm-c/tlc.h]: remove `min_p` from here, instead:
1. add it as a param to `add_request` (after `llgs`)
2. add `min_p` as a field in `ReqData`
3. add `min_p` it as a field in `PendingSeq`
4. There will be some places you need to update in `async_exec` based on steps 2 & 3. You can generally search for `out_mask_pointer` and look for the places it assigns it and do something similar to `min_p`.

Intuition for above changes is that we want to more closely model the data flow of `min_p` after that of `llgs`. All of these other parameters (e.g. `top_p`, `frequence_penalty`, etc.) are processed internally by TRT-LLM, but `llgs` is processed by the custom code within llgtrt.

File: trtllm-c/logits.cpp [hunk 7]
```diff
@@ -34,7 +34,7 @@ void* tlc_alloc_logit_data(int32_t mask_stride_, int32_t max_batch_size_)
     assert(max_batch_size > 0);
     assert(mask_stride % 4 == 0);
 
-    size_t hd_size = max_batch_size * sizeof(int64_t) * 4;
```
Comment (user-0a) [inline trtllm-c/logits.cpp:37]: In this file, where it does `entry.out_mask_pointer = nullptr;` and `entry.temperature = 1.0f;`, do something similar for min_p e.g. `entry.min_p = -FLT_MAX`

File: trtllm-c/tlc.h (hunk 6 above)
Comment (DwieDave) [reply trtllm-c/tlc.h]: When removing `min_p` in `tlc.h` it get's removed from the `TlcLogitsEntry` struct in the rust project as well and most of the set operations for `out_mask_pointer` are happening for the property on a `TlcLogitsEntry` where min_p is now missing.

Theres only one small part where the `PendingSeq` is targeted.

File: trtllm-c/logits.cpp (hunk 7 above)
Comment (DwieDave) [reply trtllm-c/logits.cpp:37]: This entry also references a `TlcLogitsEntry` struct where with above changes the min_p property would be missing.

File: trtllm-c/tlc.h [hunk 8]
```diff
@@ -23,6 +23,8 @@ extern "C"
         uint32_t _num_tokens;
         // set by the callback (initially 1.0)
         float temperature;
+        // set by the callback (initially 0.0)
+        float min_p;
```
Comment (user-0a) [inline trtllm-c/tlc.h]: Ah I think this should be `ln_min_p`. To be honest not 100% sure why the maintainer suggested to use `min_p` over `ln_min_p` in the processor (maybe to maintain high performance as the latter only requires a simple comparison), but I guess that the conversion should happen before it gets packed into this struct (similar to how the grammar is converted to an `out_mask_pointer` in this struct)

File: trtllm-c/logits.cpp [hunk 9]
```diff
@@ -109,6 +110,7 @@ static void logitsPostProcessorFn(std::vector<tle::IdType> const& reqIds, std::v
         entry._num_tokens = tokens[i].get()[0].size();
         entry.out_mask_pointer = nullptr;
         entry.temperature = 1.0f;
+        entry.min_p = -FLT_MAX;
```
Comment (user-0a) [inline trtllm-c/logits.cpp]: `ln_min_p`

File: llgtrt/src/async_exec.rs [hunk 10]
```diff
@@ -145,6 +147,7 @@ impl PendingSeq {
         let mask = step_res.sample_mask.as_ref().expect("No mask");
         self.entry.out_mask_pointer = copy_mask(mask);
         self.entry.temperature = llg.temperature;
+        self.entry.min_p = self.min_p;
```
Comment (user-0a) [inline llgtrt/src/async_exec.rs]: `min_p` -> `ln_min_p` conversion should probably happen here

File: trtllm-c/tlc.h (hunk 8 above)
Comment (user-0a) [reply trtllm-c/tlc.h]: (and update the comment above the field correspondingly)

File: trtllm-c/logits.cpp [hunk 11]
```diff
@@ -188,6 +192,7 @@ static void logitsPostProcessorFn(std::vector<tle::IdType> const& reqIds, std::v
         masksOffsets[dp] = mask_offset;
         temperatures[dp] = entries[i].temperature;
```
Comment (user-0a) [inline trtllm-c/logits.cpp:195]: I think that you need something here like `lnMinPs[dp] = entries[i].ln_min_p;`. Without it, the `min p` value is never used.
